### ISSUE TITLE
EdgeHub Disconnecting modules in 1.0.5

### ISSUE THREAD

**Issue by MPapst** [Azure/iotedge#673]
After a while of running, the edgeHub disconnects one of my C# Modules with the an exception. Only a restart of the module helps out for a couple of minutes (< 20).

## Expected Behavior
IoT Edge Module should be constantly receiving messages and should be creating new messages out of them.

## Current Behavior
The following exception occurs:
```
2019-01-02 14:03:40.886 +01:00 [WRN] - Error sending messages to module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
System.TimeoutException: Message completion response not received
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.SendMessageAsync(IMessage message, String input) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 266
   at Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint.ModuleMessageProcessor.<>c__DisplayClass5_2.<<ProcessAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/ModuleEndpoint.cs:line 99
2019-01-02 14:03:40.887 +01:00 [WRN] - Unable to send message to 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression because ModuleMessages link was not found.
2019-01-02 14:04:10.886 +01:00 [WRN] - Did not receive ack for message 72500656-7eb2-4c33-861b-a0cc4177c443 from device/module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-02 14:04:10.887 +01:00 [WRN] - Error sending messages to module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
System.TimeoutException: Message completion response not received
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.SendMessageAsync(IMessage message, String input) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 266
   at Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint.ModuleMessageProcessor.<>c__DisplayClass5_2.<<ProcessAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/ModuleEndpoint.cs:line 99
2019-01-02 14:04:10.888 +01:00 [WRN] - Unable to send message to 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression because ModuleMessages link was not found.
2019-01-02 14:04:40.887 +01:00 [WRN] - Did not receive ack for message 3aa3c575-4bd0-48f5-8a97-6a56e0b82999 from device/module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-02 14:04:40.887 +01:00 [WRN] - Error sending messages to module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
System.TimeoutException: Message completion response not received
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.SendMessageAsync(IMessage message, String input) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 266
   at Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint.ModuleMessageProcessor.<>c__DisplayClass5_2.<<ProcessAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/ModuleEndpoint.cs:line 99
2019-01-02 14:04:40.888 +01:00 [WRN] - Unable to send message to 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression because ModuleMessages link was not found.
2019-01-02 14:05:10.886 +01:00 [WRN] - Did not receive ack for message dd9d84c5-6e53-4d4a-b67e-c72d9bc5936a from device/module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-02 14:05:10.887 +01:00 [WRN] - Error sending messages to module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
System.TimeoutException: Message completion response not received
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.SendMessageAsync(IMessage message, String input) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 266
   at Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint.ModuleMessageProcessor.<>c__DisplayClass5_2.<<ProcessAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/ModuleEndpoint.cs:line 99
```
After the Exception the edgeHub shows message like these:
```
2019-01-02 15:00:15.008 +01:00 [INF] - Token updated for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/modbus
2019-01-02 15:01:08.707 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-02 15:01:25.998 +01:00 [INF] - Existing token not found for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression. Getting new token from the client...
2019-01-02 15:01:25.998 +01:00 [INF] - Connection status for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression changed to TokenNearExpiry
2019-01-02 15:01:25.998 +01:00 [INF] - Existing token not found for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression. Getting new token from the client...
2019-01-02 15:01:26.014 +01:00 [INF] - New token received on the Cbs link
2019-01-02 15:01:26.015 +01:00 [INF] - Token updated for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-02 15:01:45.999 +01:00 [INF] - Connection status for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression changed to TokenNearExpiry
2019-01-02 15:01:46.000 +01:00 [INF] - Obtained new token for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression that expires in 00:59:39.9999114
2019-01-02 15:02:08.704 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-02 15:03:08.704 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-02 15:03:10.338 +01:00 [INF] - Reauthenticating connected clients
2019-01-02 15:04:08.707 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-02 15:05:08.707 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-02 15:06:08.708 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-02 15:07:08.707 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
```

journalctl shows one ugly message:
```
Jan 02 14:21:08 svr-nar-iot iotedged[11981]: 2019-01-02T14:21:08Z [ERR!] - server connection error: (unknown)
Jan 02 14:21:08 svr-nar-iot iotedged[11981]: 2019-01-02T14:21:08Z [ERR!] - error writing a body to connection: Broken pipe (os error 32)
```

## Steps to Reproduce
I am running 2 modules (except Hub and Agent) with two routes.

* Modbus to Compression
* Compression to Upstream

The error occurs since the update to 1.0.5.

### Device (Host) Operating System
Ubuntu 18.04 VM on Hyper-V

### Architecture
amd64

### Container Operating System
Linux Containers

#### iotedged
`iotedge 1.0.5 (d76e0316c6f324345d77c48a83ce836d09392699)`

#### Edge Agent
`2019-01-02 10:07:54.236 +00:00 [INF] - Version - 1.0.5.19141174 (d76e0316c6f324345d77c48a83ce836d09392699)`

#### Edge Hub
`2019-01-02 11:08:06.028 +01:00 [INF] - Version - 1.0.5.19141174 (d76e0316c6f324345d77c48a83ce836d09392699)`

#### Docker
```
iotadmin@svr-nar-iot:~$ sudo docker version
Client:
 Version:           3.0.2
 API version:       1.40
 Go version:        go1.11.1
 Git commit:        e042b58f
 Built:             Tue Dec  4 20:13:19 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          3.0.2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.11.2
  Git commit:       74baf62
  Built:            Tue Dec  4 20:29:24 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

## Logs

see above

## Additional Information

Device SDK Version is 1.18.1, .net core Runtime Version is 2.1, all modules are connected using AMQP

**Comment by MPapst** [comment]
Question: Does the exception message `Message completion response not received` mean that the module is taking to long for working on the message?
-> If yes, what is the timeout?
-> If yes, why does it completely disconnect the module, the module itself does not show any exception
-> If no, what does it mean?

**Comment by varunpuranik** [comment]
@MPapst - Thanks for reporting the issue. 
From the logs you shared, it seems like the EdgeHub is able to get a token from the module and keep the connection but at the same time it reports that the connection to the module was not found. This is not the expected behavior and could be a bug in EdgeHub. We will investigate. 
Meanwhile, as a temporary workaround, can you try switching the compression module to use MQTT instead of AMQP, and see if that fixes it?

To answer your other question - the message "Message completion response not received" does mean that the ack for the message was not received by the EdgeHub in time, so it tries to resend the message. The timeout for that is 30secs. However, the edgeHub will not disconnect the module connection when that happens.

**Comment by MPapst** [comment]
Hi @varunpuranik, 
thanks for the reply.

I just changed the connection (only for the compression module) to MQTT.
As the compression module is the upstream route for the other module(s), I am not getting any messages in the cloud when this one does not have a hub connection --> There has been a spike in the message the device to cloud messages directly after the deployment of the MQTT connected module and it stopped directly afterwards:
![grafik](https://user-images.githubusercontent.com/16494676/50626368-11058f00-0f2e-11e9-8a49-37e46d1af71a.png)

```
2019-01-03 07:48:12.375 +01:00 [INF] - Authenticated 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with IotHub
2019-01-03 07:48:12.375 +01:00 [INF] - Successfully generated identity for clientId 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression and username svr-nar-iot/978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression/?api-version=2018-06-30&DeviceClientType=Microsoft.Azure.Devices.Client%2F1.18.1%20%28.NET%20Core%204.6.27019.06%3B%20Linux%204.15.0-43-generic%20%2346-Ubuntu%20SMP%20Thu%20Dec%206%2014%3A45%3A28%20UTC%202018%3B%20X64%29
2019-01-03 07:48:12.375 +01:00 [INF] - ClientAuthenticated, 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression, 1a9fc75c
2019-01-03 07:48:12.378 +01:00 [INF] - New device connection for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:48:12.378 +01:00 [INF] - Bind device proxy for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:48:12.378 +01:00 [INF] - Binding message channel for device Id 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:48:12.381 +01:00 [INF] - Processing subscription TwinResponse for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.
2019-01-03 07:48:12.381 +01:00 [INF] - Processing subscription Methods for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.
2019-01-03 07:48:12.465 +01:00 [INF] - Processing subscription ModuleMessages for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.
2019-01-03 07:48:12.465 +01:00 [INF] - Set subscriptions from session state for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:48:12.642 +01:00 [INF] - Started task to cleanup processed and stale messages for endpoint iothub
2019-01-03 07:48:14.455 +01:00 [INF] - Cloud connection for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is True
2019-01-03 07:48:14.455 +01:00 [INF] - Connection status for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression changed to ConnectionEstablished
2019-01-03 07:48:14.473 +01:00 [INF] - Cleaned up 5325 messages from queue for endpoint iothub and 5325 messages from message store.
2019-01-03 07:53:10.340 +01:00 [INF] - Reauthenticating connected clients
2019-01-03 07:53:10.340 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with cached service identity 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression. Resyncing service identity...
2019-01-03 07:53:10.624 +01:00 [WRN] - Received null device scope result for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:10.633 +01:00 [INF] - Service identity for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression was removed from device scope, dropping client connection.
2019-01-03 07:53:10.634 +01:00 [WRN] - Closing connection for device: 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression, Microsoft.Azure.Devices.Edge.Hub.Core.EdgeHubConnectionException: Connection closed for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.,
2019-01-03 07:53:10.634 +01:00 [INF] - Disposing MessagingServiceClient for device Id 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression because of exception - Microsoft.Azure.Devices.Edge.Hub.Core.EdgeHubConnectionException: Connection closed for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.
2019-01-03 07:53:10.634 +01:00 [INF] - Closing device proxy for device Id 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:10.634 +01:00 [INF] - Device connection removed for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:10.634 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with module credentials. Attempting to authenticate using device 978f66f0-d43e-4ff6-92bb-0b0895c13a06 credentials.
2019-01-03 07:53:10.634 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with cached service identity 978f66f0-d43e-4ff6-92bb-0b0895c13a06. Resyncing service identity...
2019-01-03 07:53:10.886 +01:00 [WRN] - Received null device scope result for 978f66f0-d43e-4ff6-92bb-0b0895c13a06
2019-01-03 07:53:10.896 +01:00 [INF] - Service identity for 978f66f0-d43e-4ff6-92bb-0b0895c13a06 was removed from device scope, dropping client connection.
2019-01-03 07:53:11.056 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-03 07:53:11.669 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with cached service identity 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression. Resyncing service identity...
2019-01-03 07:53:11.915 +01:00 [WRN] - Received null device scope result for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:11.925 +01:00 [INF] - Service identity for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression was removed from device scope, dropping client connection.
2019-01-03 07:53:11.925 +01:00 [INF] - Device connection removed for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:11.926 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with module credentials. Attempting to authenticate using device 978f66f0-d43e-4ff6-92bb-0b0895c13a06 credentials.
2019-01-03 07:53:11.926 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with cached service identity 978f66f0-d43e-4ff6-92bb-0b0895c13a06. Resyncing service identity...
2019-01-03 07:53:12.055 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-03 07:53:12.205 +01:00 [WRN] - Received null device scope result for 978f66f0-d43e-4ff6-92bb-0b0895c13a06
2019-01-03 07:53:12.217 +01:00 [INF] - Service identity for 978f66f0-d43e-4ff6-92bb-0b0895c13a06 was removed from device scope, dropping client connection.
2019-01-03 07:53:12.217 +01:00 [INF] - Client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression in device scope not authenticated locally.
2019-01-03 07:53:12.217 +01:00 [INF] - Attempting to connect to IoT Hub for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression via AMQP...
2019-01-03 07:53:12.218 +01:00 [INF] - New token requested by client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression, but using existing token as it is usable.
2019-01-03 07:53:12.302 +01:00 [INF] - Cloud connection for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is True
2019-01-03 07:53:12.302 +01:00 [INF] - Created cloud proxy for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression via AMQP, with client operation timeout 20 seconds.
2019-01-03 07:53:12.302 +01:00 [INF] - Initialized cloud proxy 1803fcc6-58ad-4ef2-a862-cbf72a2c2dd4 for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:12.303 +01:00 [INF] - Closing receiver in cloud proxy 371c13b2-9f6f-49d9-8f8e-23fb29658d35 for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:12.303 +01:00 [INF] - Closed cloud proxy 371c13b2-9f6f-49d9-8f8e-23fb29658d35 for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:12.303 +01:00 [INF] - New cloud connection created for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:12.303 +01:00 [INF] - Authenticated 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with IotHub
2019-01-03 07:53:12.303 +01:00 [INF] - Successfully generated identity for clientId 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression and username svr-nar-iot/978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression/?api-version=2018-06-30&DeviceClientType=Microsoft.Azure.Devices.Client%2F1.18.1%20%28.NET%20Core%204.6.27019.06%3B%20Linux%204.15.0-43-generic%20%2346-Ubuntu%20SMP%20Thu%20Dec%206%2014%3A45%3A28%20UTC%202018%3B%20X64%29
2019-01-03 07:53:12.303 +01:00 [INF] - ClientAuthenticated, 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression, 0cbbe0b4
2019-01-03 07:53:12.304 +01:00 [INF] - New device connection for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:12.305 +01:00 [INF] - Bind device proxy for device 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:12.305 +01:00 [INF] - Binding message channel for device Id 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-03 07:53:12.306 +01:00 [INF] - Processing subscription TwinResponse for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.
2019-01-03 07:53:12.306 +01:00 [INF] - Processing subscription Methods for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.
2019-01-03 07:53:12.396 +01:00 [INF] - Processing subscription ModuleMessages for client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression.
2019-01-03 07:53:12.397 +01:00 [INF] - Set subscriptions from session state for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
```

**Comment by MPapst** [comment]
While having issues with MQTT also (right now the module is disconnected after a couple of hours), today I got this after restarting the compression module:

```
System.Threading.Tasks.TaskCanceledException: The operation was canceled. ---> System.IO.IOException: Unable to read data from the transport connection: Operation canceled. ---> System.Net.Sockets.SocketException: Operation canceled
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)
   at System.Net.Security.SslStreamInternal.<FillBufferAsync>g__InternalFillBufferAsync|38_0[TReadAdapter](TReadAdapter adap, ValueTask`1 task, Int32 min, Int32 initial)
   at System.Net.Security.SslStreamInternal.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.DeviceScopeApiClient.GetIdentitiesInScope(Uri uri) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/DeviceScopeApiClient.cs:line 105
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.DeviceScopeApiClient.GetIdentitiesInScopeWithRetry(Uri uri) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/DeviceScopeApiClient.cs:line 85
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ServiceProxy.GetServiceIdentity(String deviceId) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ServiceProxy.cs:line 30
   at Microsoft.Azure.Devices.Edge.Hub.Core.DeviceScopeIdentitiesCache.RefreshServiceIdentity(String id) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs:line 140
```
This is what the edgeHub logs are showing when the module is no longer connected:
```
2019-01-08 11:08:14.068 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-08 11:08:17.740 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-08 11:08:24.689 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected
2019-01-08 11:08:25.257 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with cached service identity 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression. Resyncing service identity...
2019-01-08 11:08:25.514 +01:00 [WRN] - Received null device scope result for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
2019-01-08 11:08:25.524 +01:00 [INF] - Service identity for 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression was removed from device scope, dropping client connection.
2019-01-08 11:08:25.525 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with module credentials. Attempting to authenticate using device 978f66f0-d43e-4ff6-92bb-0b0895c13a06 credentials.
2019-01-08 11:08:25.525 +01:00 [INF] - Unable to authenticate client 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression with cached service identity 978f66f0-d43e-4ff6-92bb-0b0895c13a06. Resyncing service identity...
2019-01-08 11:08:42.463 +01:00 [WRN] - Module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression is not connected

```

**Comment by varunpuranik** [comment]
Hi @MPapst - Looks like you are running into a bug in our EdgeHub, the fix for which just got merged to master - https://github.com/Azure/iotedge/pull/684
This fix will be available in our next release (currently planned for the end of this month). 
To work around this issue, can you create a new Edge device in your IoTHub and use that instead? That should cause the problem to go away.

**Comment by MPapst** [comment]
Still there, even with new device and new iothub.

Now I'm, getting this exception even with MQTT:
```
2019-01-09 14:39:00.018 +01:00 [WRN] - Error sending messages to module 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression
System.TimeoutException: Message completion response not received
   at Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler.SendMessageAsync(IMessage message, String input) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/device/DeviceMessageHandler.cs:line 266
   at Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint.ModuleMessageProcessor.<>c__DisplayClass5_2.<<ProcessAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/ModuleEndpoint.cs:line 99
2019-01-09 14:42:31.026 +01:00 [INF] - Closing connection for device: 978f66f0-d43e-4ff6-92bb-0b0895c13a06/compression, ,

```

**Comment by varunpuranik** [comment]
I see interesting, maybe you are running into some other issue as well then. Can you please share more EdgeHub logs with the new device? It is hard to tell what is going on from the small snippet. 
I appreciate your patience with this. Hopefully we will be able to sort it out soon.

**Comment by MPapst** [comment]
Please find it attached: [edgeHub.txt](https://github.com/Azure/iotedge/files/2742617/edgeHub.txt)
This is the current log since the new device registration.

**Comment by MPapst** [comment]
intermediate fix: Created a cronjob restarting the module every hour.

**Comment by varunpuranik** [comment]
I looked at the logs - I don't see anything in the logs that jumps out to me as odd or something to indicate that the system is not working correctly. The EdgeHub disconnects the module compression after the token expires, which is by design, but the module seems to be reconnecting, again as expected. 
If the module doesn't reconnect intermittently (though I don't see that in the logs), then that might be an SDK issue. If you can confirm that, then we can open an issue on the SDK.

**Comment by MPapst** [comment]
I'll open a new SDK issue.

Thanks!

**Comment by MPapst** [comment]
I think this could be fixed with 1.19 of the SDK (as per Azure/azure-iot-sdk-csharp#558). I will try that one and open a ticket after that.
